### PR TITLE
Usds attack

### PR DESCRIPTION
**Pull request by nine-december**
This should be ready to go!

### CODE REVIEW

File: test/Business_Logic/Usds/README.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    function isContract(address account) internal view returns (bool) {
+        // This method relies on extcodesize/address.code.length, which returns 0
+        // for contracts in construction, since the code is only stored at the end
+        // of the constructor execution.
+
+        return account.code.length > 0;
+    }
+```
+
+This implementation was before performed by using the low level `extcodesize` call, now checking directly the `account.code.length` property. If `isContract` returns false, it does not assures that the address checked won't have deployed code across its lifecycle. Some examples that return false when calling `isContract`:
```
Comment (joaquinlpereyra) [inline test/Business_Logic/Usds/README.md:48]: `This implementation was before performed by using the low level `extcodesize` call, now checking directly the `account.code.length` property.` 

How is this relevant?

File: test/Business_Logic/Usds/README.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    }
+```
+
+The address bypass allows the attacker to generate balance in a 'non contract' account. Later, once the contract is deployed, the first transfer assigns the attackers `nonRebasingCreditsPerToken[_account] = 1` inside `_ensureRebasingMigration()`. Then, when calculating its balance it does not divide the amount of `credits` by `rebasingCreditsPerToken` and simply returns the amount of `credits` without rebasing.
+
+The attacker managed to swap an equivalent of ~309K USD.
+
+## Possible mitigations
+
+1.  Never rely on `isContract` or similar to check that an address will never be a contract across it's lifecycle.
```
Comment (joaquinlpereyra) [inline test/Business_Logic/Usds/README.md:98]: `s/'//`

File: test/Business_Logic/Usds/usds.attack.sol [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+contract Exploit_Usds is TestHarness, TokenBalanceTracker {
+    IERC20 internal usds = IERC20(0xD74f5255D557944cf7Dd0E45FF521520002D5748); // sperax token
+    address internal usdsWhale = 0x50450351517117Cb58189edBa6bbaD6284D45902;
+
+    ContractFactory internal factory;
+
+    function setUp() external {
+        cheat.createSelectFork("arbitrum", 57803396); 
+        // deal(address(usds), address(this), 20e18); // because USDS is behind a proxy, this fails on foundry.
```
Comment (joaquinlpereyra) [inline test/Business_Logic/Usds/usds.attack.sol:18]: OK, fails, what's the alternative? Or is it not needed here?

**Review by joaquinlpereyra — COMMENTED** [review]
Wouldn't it be cool to use the actual Gnosis-Safe Factory like the attacker did?

File: test/Business_Logic/Usds/README.md (hunk 1 above)
Comment (nine-december) [reply test/Business_Logic/Usds/README.md:48]: Not that relevant, just depicting two ways of achieving the same.

File: test/Business_Logic/Usds/usds.attack.sol (hunk 3 above)
Comment (nine-december) [reply test/Business_Logic/Usds/usds.attack.sol:18]: The alternative, we use the whale pranking method starting from line 21.

Comment (joaquinlpereyra) [reply test/Business_Logic/Usds/usds.attack.sol:18]: Should be better explained that that the prank is needed because of this.

**Review by nine-december — COMMENTED** [review]
Thanks for the improvements!

**Review by joaquinlpereyra — APPROVED** [review]

**Pull request merged**